swift/ql/lib/codeql/swift/elements/Element.qll

```diff
@@ -1,4 +1,11 @@
-// generated by codegen/codegen.py, remove this comment if you wish to edit this file
 private import codeql.swift.generated.Element
 
-class Element extends ElementBase { }
+class Element extends ElementBase {
```

Comment:
Something that still eludes me. Why is there both an `Element` and an `ElementBase`? Why isn't there just one class?

Reply:
There are a couple of autogenerated predicates whose rootdef is in `ElementBase` (e.g., the `resolve` predicate that enables us to hide implicit conversions from the "main" AST ). We _could_ put those predicates in `Element` instead, but then we'd have a weird situation where `Element` should contain both autogenerated predicates and manually written predicates.

Reply:
So the general pattern will be: each `Base` class only contains auto generated stuff, and each class extending it will contain handwritten functionality?

Reply:
You might find more details about this choice in [this section of the EDR document](https://docs.google.com/document/d/1vFVZKEvocpQE-9SA1EjYfhKAIyTNGCxxdloIAZWdkvY/edit#heading=h.sinlay1iaclz).

In a nutshell, this is the separation we have between fully generated methods and hand written ones. `XXXBase` is fully generated and kept up to date with `schema.yml`. `XXX` is initially a generated empty stub, but can be customized inheriting from the generated one. 

The alternatives could be:
* one generated file, that stops being generated once you edit it -> this would mean you would lose new generated methods and would need to import them by hand
* like now with two files, but without using inheritance -> you would probably need a lot of boilerplate to "import" the generated implementations in the non-generated file (I think this is similar to what python has)
* some kind of specifications of what needs to be added to the generated file (a diff file? some other "include" specification for the codegen suite? probably something triggered from `schema.yml` or some kind of annotation file thereof): seemed a less clean solution requiring more effort

Reply:
Current approach is fine with me. I was just trying to understand what is going on.